src/Npgsql/NpgsqlConnector.cs

```diff
@@ -1129,8 +1150,22 @@ internal ValueTask<IBackendMessage> ReadMessage(bool async, DataRowLoadingMode d
                     {
                         EndUserAction();
                     }
+
+                    // We failed because of a timeout, and the cancellation was successful
+                    if (!(originalTimeoutException is null) && e.SqlState == PostgresErrorCodes.QueryCanceled)
+                    {
+                        var tempException = originalTimeoutException;
+                        originalTimeoutException = null;
+                        throw tempException;
+                    }
+
                     throw;
                 }
+                catch (NpgsqlException e) when (!readingNotifications2 && e.InnerException is TimeoutException && !(originalTimeoutException is null))
```

Comment:
The code would probably be easier to follow/less duplicated if this logic were simply included into the timeout catch you already added above. So instead if having one timeout catch above for the first timeout, and another one here for the second timeout, you could have a single catch above, which internally check originalTimeoutException and branches on it.